### PR TITLE
Changes to API Auth

### PR DESCRIPTION
**Pull request by tcitworld**
Therefore, username and password are no longer needed

| Q             | A
| ------------- | ---
| Bug fix?      | yes
| New feature?  | no
| BC breaks?    | yes
| Deprecations? | yes
| Tests pass?   | yes
| Documentation | yes
| Translation   | no
| Fixed tickets | #2800 #3005
| License       | MIT

I'll like to put in more things in this PR, but feel free to review and test.

- [x] Allow admin to create global clients (not user related) for automatic login for wallabag.it / framabag users in apps. This will need the `/oauth/v2/auth` route and the `authorization_code` `grant_type`.
- [x] Make `redirect_uri` not required on simple client creation.
- [x] Add back `password` `grant_type` but depreciate it.
- [ ] Tests
- [x] Translations
- [x] Better authorize page
- [ ] Database migration

Ref. https://github.com/wallabag/wallabag/issues/2800#issuecomment-286679410

### CODE REVIEW

**Comment by tcitworld** [comment]
### User specific clients
Users can create user-specific clients, which gives them a `client_id` and `client_password` (like before), which any can app use with the `client_credentials` `grant_type` to get directly an `access_token`, **without providing username and password like before**. Callback URIs are no longer required.

### Global clients
Instances admin can create global clients accessible for all of the instance users, with a name, one or multiple callback URIs, a description (**new!**) and a picture (**new!**). Then they must contact the app maintainers to add their instance with the corresponding client_id.
There's now two methods to connect, depending if the app is server-side (or can obfuscate client secrets) or if it's client-side.
#### Server-side
Applications first call `https://wallabag.tld/oauth/v2/auth` with the following parameters : 
  * `client_id`: the one the instance administrator gave
  * `redirect_uri`: the uri (schemes allowed !) to redirect after authorization is made, read below)
  * `response_type`: `code`
  * `scope`: There is `read`, `write` and `user` (ref #3065)

Then, wallabag promps a page (after login if needed) asking the user to allow the app and then redirects to the `redirect_uri` with a code as parameter.
The app now calls `https://wallabag.tld/oauth/v2/auth` with the following parameters : 
  * `code`: the code you just got
  * `client_id`: the client id, same as before
  * `redirect_uri`: the final redirect uri
  * `grant_type`: `authorization_code`

The `redirect_uri` is finally called with the `access_token` (and `refresh_token` and others like before...) as parameter.

#### Client-side
Applications first call `https://wallabag.tld/oauth/v2/auth` with the following parameters : 
  * `client_id`: the one the instance administrator gave
  * `redirect_uri`: the uri (schemes allowed !) to redirect after authorization is made, read below)
  * `response_type`: `token`
  * `scope`: There is `read`, `write` and `user` (ref #3065)

The `redirect_uri` is then called with the `access_token` (and `refresh_token` and others like before...) as parameter.

**Comment by tcitworld** [comment]
Ping @wallabag/contributors @wallabag/core ^

File: src/Wallabag/ApiBundle/Controller/DeveloperController.php [hunk 1]
```diff
@@ -20,9 +22,62 @@ class DeveloperController extends Controller
     public function indexAction()
     {
         $clients = $this->getDoctrine()->getRepository('WallabagApiBundle:Client')->findByUser($this->getUser()->getId());
+        $global_clients = $this->getDoctrine()->getRepository('WallabagApiBundle:Client')->findByUser(null);
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/DeveloperController.php]: camelCase please ➡️ `$globalClients`

File: src/Wallabag/ApiBundle/Controller/DeveloperController.php [hunk 2]
```diff
@@ -75,10 +130,16 @@ public function createClientAction(Request $request)
      */
     public function deleteClientAction(Client $client)
     {
-        if (null === $this->getUser() || $client->getUser()->getId() != $this->getUser()->getId()) {
+        if (($client->getUser() !== null) && (null === $this->getUser() || $client->getUser()->getId() != $this->getUser()->getId())) {
             throw $this->createAccessDeniedException('You can not access this client.');
         }
 
+        if ($img = $client->getImage()) {
+            if ($file = $this->getParameter('wallabag_api.applications_icon_path') . '/' . $img) {
```
Comment (j0k3r) [inline src/Wallabag/ApiBundle/Controller/DeveloperController.php]: You can merge these both condition

**Comment by matteocoder** [comment]
@tcitworld should the whole comment be translated or should I leave something out (like the content inside the parentheses)?

**Comment by tcitworld** [comment]
@matteocoder Which comment ? I haven't started the translations.

File: src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml [hunk 3]
```diff
@@ -445,13 +445,17 @@ developer:
     clients:
         title: "Clients"
         create_new: "Créer un nouveau client"
+        create_new_global: Créer un nouveau client global
```
Comment (nicosomb) [inline src/Wallabag/CoreBundle/Resources/translations/messages.fr.yml]: Maybe you can add quotes for each translation.

File: app/Resources/FOSOAuthServerBundle/views/Authorize/authorize_content.html.twig [hunk 4]
```diff
@@ -0,0 +1,52 @@
+<div class="row">
```
Comment (nicosomb) [inline app/Resources/FOSOAuthServerBundle/views/Authorize/authorize_content.html.twig]: Don't forget translations for this file.

**Comment by matteocoder** [comment]
@tcitworld sorry, I got a little confused. You referenced this PR in another one, so I though that some other text needed to be translated.

File: src/Wallabag/ApiBundle/Controller/DeveloperController.php [hunk 5]
```diff
@@ -20,9 +22,62 @@ class DeveloperController extends Controller
...
+            $client->setAllowedGrantTypes(['token', 'authorization_code', 'refresh_token']);
+            $em->persist($client);
+            $em->flush();
+
+            $this->get('session')->getFlashBag()->add(
+                'notice',
+                $this->get('translator')->trans('flashes.developer.notice.client_created', ['%name%' => $client->getName()])
+            );
+
+            return $this->render('@WallabagCore/themes/common/Developer/client_parameters.html.twig', [
```
Comment (nicosomb) [inline src/Wallabag/ApiBundle/Controller/DeveloperController.php]: Maybe you can display the uploaded picture when we just have created the client.

**Comment by jlnostr** [comment]
@tcitworld Call me stupid, but I don't understand the new behavior as there are several questions open.

1. How should I call the given URL? Inside a webbrowser visible for the user or hidden?
2. How is the redirect URI called with the tokens? Is there a scheme for it? How will it connect with the given redirect URI?

It would be nice to clarify the auth behavior and a recommendation which (server or client) behavior an app should use.

**Comment by jlnostr** [comment]
After this PR the QR codes for setting up a client should also be updated.

**Comment by tcitworld** [comment]
> How should I call the given URL? Inside a webbrowser visible for the user or hidden?

Yes, the user must interact to login then allow your app.

> How is the redirect URI called with the tokens?

The redirect URI is called with tokens as `GET` parameters.

Is there a scheme for it? How will it connect with the given redirect URI?

I don't get these points.

**Comment by jlnostr** [comment]
Forget my last point. Assuming I have the following redirect URI:

    wallabag://login_success

How is the final result going to look like? Like this?

    wallabag://login_success?access_token=abcd&refresh_token=1234

**Comment by tcitworld** [comment]
That's right.

**Comment by Simounet** [comment]
I didn't do this part on wallabagger but if I get it well:
- there is no more `oauth/v2/token` but `oauth/v2/auth`
- the user need to interact directly with wallabag (on a popup for example)
- we need to get and store the result of the previous page mentioned

**Comment by tcitworld** [comment]
Okay so I've changed it a bit. I found that admins registering themselves the apps was a pain in the ass so now apps can register themselves by sending a POST request at `/api/apps` with the following parameters : 

* `client_name`: **REQUIRED** the client name, typically your app name, that will be shown to the user
* `redirect_uri`: **REQUIRED** the redirect uris used through the auth process
* `logo_uri`: An URL for your logo that will be displayed on the allow page
* `description`: A short description for your app that will be displayed at the user
* `app_uri`: An URL for your app, it's homepage or help section...

In other changes, the access_token delivered how has a lifetime of **60 days** and refresh_tokens are no longer expired. You also need to pass a state (random string) when requiring permission from the user on `/oauth/v2/auth` and check it afterwards.

Docs incoming. Translations now welcomed.

**Comment by tcitworld** [comment]
> there is no more oauth/v2/token but oauth/v2/auth

No. `/oauth/v2/auth` is used to allow the user to accept the app to use his account. `/oauth/v2/token` is only used in the Server-side to exchange a code for an `access_token`. Otherwise, `/oauth/v2/auth` returns directly the `access_token` in the redirect uri callback.
> the user need to interact directly with wallabag (on a popup for example)

Nope, you need to register a callback to this kind of URIs : https://developer.chrome.com/apps/identity#method-launchWebAuthFlow

> we need to get and store the result of the previous page mentioned

Hmm ?

**Comment by tcitworld** [comment]
Also, added a « Featured apps page » (added wallabagger afterwards).
wallabag_prez

**Comment by jlnostr** [comment]
> now apps can register themselves by sending a POST request
And this doesn't need any kind of auth? If so, then we should worry about a lot of spam clients because everyone can create a dozen of useless clients.

Another question: With the new `oauth/v2/auth` way, refreshing the tokens is only necessary once a year, right? And it still uses the same "old" API endpoints? Or is refreshing not longer needed?

**Comment by tcitworld** [comment]
> And this doesn't need any kind of auth? If so, then we should worry about a lot of spam clients because everyone can create a dozen of useless clients.

Yes, but I plan to have a task/command (called by a cron for instance) to remove all useless clients that didn't get an approval from the user.

> Another question: With the new oauth/v2/auth way, refreshing the tokens is only necessary once a year, right? And it still uses the same "old" API endpoints? Or is refreshing not longer needed?

That's what I was hoping for but the library we use makes refreshing required. So refreshing the tokens is needed once a year. The `/oauth/v2/token` route stays the same.

**Comment by di72nn** [comment]
>> And this doesn't need any kind of auth? If so, then we should worry about a lot of spam clients because everyone can create a dozen of useless clients.

> Yes, but I plan to have a task/command (called by a cron for instance) to remove all useless clients that didn't get an approval from the user.

That's not acceptable. I'm yet to read through the changes to suggest anything.

**Comment by tcitworld** [comment]
> That's not acceptable

Please develop.

**Comment by di72nn** [comment]
Allowing to perform not read-only actions on a server without any authentication is an easy way to introduce DoS or other vulnerabilities.

> a task/command (called by a cron for instance) to remove all useless clients that didn't get an approval from the user

That doesn't sound like a viable solution.


IMO, user-specific apps should be added with at least user access level, global apps should be added with admin user access level. A flow like `wallabag promps a page (after login if needed) asking the user to allow the app` should be fine I guess.

**Comment by di72nn** [comment]
Also I don't see the point in overly increasing tokens' lifetime considering improved auth mechanism.

**Comment by tcitworld** [comment]
The idea is for apps to login users with a good user experience. It shouldn't be needed to create client credentials on the server or even scan a QR code before using them.
On big instances the admins creating the global apps could be fine, but most of our users aren't here anyway.

Thoughts @wallabag/core ?

> Also I don't see the point in overly increasing tokens' lifetime considering improved auth mechanism.
 
That's not a improvement, just a reasonable adjustement for web apps which can't get a `refresh_token` since they use the implicit flow.

**Comment by di72nn** [comment]
> The idea is for apps to login users with a good user experience. It shouldn't be needed to create client credentials on the server or even scan a QR code before using them.

 * Leave the same initial call to `/api/apps`.
 * Require a user to login (if he is not logged in already) before adding a new app.
  Isn't the user supposed to login on the next step anyway (when he authorizes a device for the app)?
 * If the app doesn't exist ask user to create it (simple OK/cancel form). If the app is already registered, return existing app info.
 * Proceed with auth.

In this scenario the `/api/apps` method allows to get credentials for existing or new apps to avoid hassle with QR code scanning (or whatever) and avoids creating duplicates of the same app.

**Comment by tcitworld** [comment]
Doable.

**Comment by tcitworld** [comment]
If I understand correctly.

* the app calls the `api/apps` route with the following parameters : 
   * `client_name`: **REQUIRED**
   * `redirect_uris`**REQUIRED**
   * `logo_uri`
   * `description`
   * `app_uri`
   * `uri_redirect_after_creation`: **NEW** - **REQUIRED** - Callback to transmit informations. *Please suggest a better parameter name*
* The user login if needed
* If the app name already exists, wallabag redirects to callback `uri_redirect_after_creation` with it's parameters
* wallabag asks the user if he wants to create the app
   * ok : callback to `uri_redirect_after_creation` with the new parameters
   * no : callback to `uri_redirect_after_creation` with an error code

However, I'm not sure the `client_name` is enough to identify an app. For instance, a fake app could give the same name and get the credentials of the previous app.
Maybe the best would be to offer the existing app or offer to create a new one, no ? @di72nn

**Comment by di72nn** [comment]
> For instance, a fake app could give the same name and get the credentials of the previous app.

The credentials don't give any access on their own, they are only used to bind clients to specific apps, right?

> Maybe the best would be to offer the existing app or offer to create a new one, no ?

Maybe add a screen that shows details about an existing app (all the fancy stuff: logo, description) with a big "Continue" button, and a subtle "Create new app instead" link/button?

**Comment by tcitworld** [comment]
> The credentials don't give any access on their own, they are only used to bind clients to specific apps, right?

No, but they could ask the user for access, and since they look the same as the real apps, users could give them access.

> Maybe add a screen that shows details about an existing app (all the fancy stuff: logo, description) with a big "Continue" button, and a subtle "Create new app instead" link/button?

Yup.

> Also, I'm not sure what's the purpose of this parameter:

>> redirect_urisREQUIRED

These are an array of the URIs used during the auth process, as the standard describes it.

**Comment by tcitworld** [comment]
I'll let you guys decide if you want this in 2.4 or not

**Pull request closed without merging**